### PR TITLE
Initial partition-aware changes for hash join.

### PR DESCRIPTION
**Pull request by adalbertgerald**
I have made some initial changes to the HashJoinOperator.

It seems that the HashJoin code has changed a lot from what we had for before, when Harshad and I worked on making it partition-aware for the quickstep group paper.

So, I would like to get some initial feedback from @hbdeshmukh and @zuyu before making any more changes.

NOTE: I haven't updated the tests yet and so the travis builds would fail.

### CODE REVIEW

File: query_optimizer/ExecutionGenerator.cpp [hunk 1]
```diff
@@ -670,10 +670,18 @@ void ExecutionGenerator::convertHashJoin(const P::HashJoinPtr &physical_plan) {
   }
 
   // Create join hash table proto.
-  const QueryContext::join_hash_table_id join_hash_table_index =
-      query_context_proto_->join_hash_tables_size();
+  std::vector<QueryContext::join_hash_table_id> join_hash_table_index;
```
Comment (zuyu) [inline query_optimizer/ExecutionGenerator.cpp]: Please rename to `join_hash_table_indexes` or `join_hash_table_indices`.

File: query_optimizer/ExecutionGenerator.cpp [hunk 2]
```diff
@@ -670,10 +670,18 @@ void ExecutionGenerator::convertHashJoin(const P::HashJoinPtr &physical_plan) {
...
+
   S::HashTable *hash_table_proto = query_context_proto_->add_join_hash_tables();
 
+  const PartitionScheme &part_scheme = build_relation_info->relation->getPartitionScheme();
+  const std::size_t num_partitions = part_scheme.getPartitionSchemeHeader().getNumPartitions();
```
Comment (zuyu) [inline query_optimizer/ExecutionGenerator.cpp]: We need to check whether `partition_scheme` exists using `hasPartitionScheme` before using it. Otherwise, we will encounter segfault.

File: query_optimizer/ExecutionGenerator.cpp [hunk 3]
```diff
@@ -698,7 +706,7 @@ void ExecutionGenerator::convertHashJoin(const P::HashJoinPtr &physical_plan) {
               build_relation_info->isStoredRelation(),
               build_attribute_ids,
               any_build_attributes_nullable,
-              join_hash_table_index));
+              join_hash_table_index[0]));
```
Comment (zuyu) [inline query_optimizer/ExecutionGenerator.cpp]: Could you please explain why we pass only the first hash table? Thanks!

File: relational_operators/HashJoinOperator.hpp [hunk 4]
```diff
@@ -120,7 +125,7 @@ class HashJoinOperator : public RelationalOperator {
                    const bool any_join_key_attributes_nullable,
                    const CatalogRelation &output_relation,
                    const QueryContext::insert_destination_id output_destination_index,
-                   const QueryContext::join_hash_table_id hash_table_index,
+                   const std::vector<QueryContext::join_hash_table_id> &hash_table_index,
```
Comment (zuyu) [inline relational_operators/HashJoinOperator.hpp]: We could use the rvalue `&&` here.

File: relational_operators/HashJoinOperator.hpp [hunk 5]
```diff
@@ -188,6 +224,21 @@ class HashJoinOperator : public RelationalOperator {
...
+  template <class JoinWorkOrderClass>
+  void addWorkOrders(WorkOrdersContainer *container,
+                     QueryContext *query_context,
+                     StorageManager *storage_manager,
+                     const Predicate *predicate,
+                     const std::vector<std::unique_ptr<const Scalar>> &selection,
+                     InsertDestination *output_destination);
+
+  template <class JoinWorkOrderClass>
+  void addPartitionAwareWorkOrders(WorkOrdersContainer *container,
```
Comment (zuyu) [inline relational_operators/HashJoinOperator.hpp]: Please be consistent when using the scope of `QUICKSTEP_HAVE_LIBNUMA`.

File: relational_operators/HashJoinOperator.hpp [hunk 6]
```diff
@@ -205,16 +256,23 @@ class HashJoinOperator : public RelationalOperator {
   const bool any_join_key_attributes_nullable_;
   const CatalogRelation &output_relation_;
   const QueryContext::insert_destination_id output_destination_index_;
-  const QueryContext::join_hash_table_id hash_table_index_;
+  std::vector<QueryContext::join_hash_table_id> hash_table_index_;
```
Comment (zuyu) [inline relational_operators/HashJoinOperator.hpp]: I believe we could still mark `const` here. Also, please rename to `hash_table_indexes_` or `hash_table_indices_`. Thanks!

File: relational_operators/HashJoinOperator.hpp [hunk 7]
```diff
@@ -265,7 +324,9 @@ class HashInnerJoinWorkOrder : public WorkOrder {
         selection_(selection),
         hash_table_(hash_table),
         output_destination_(DCHECK_NOTNULL(output_destination)),
-        storage_manager_(DCHECK_NOTNULL(storage_manager)) {}
+        storage_manager_(DCHECK_NOTNULL(storage_manager)) {
+    preferred_numa_nodes_.push_back(numa_node);
```
Comment (zuyu) [inline relational_operators/HashJoinOperator.hpp]: Question: in a machine w/o `libnuma`, what does `0` mean in `preferred_numa_nodes_`? Thanks.

**Comment by zuyu** [comment]
As of now, I have above comments. Thanks!

File: relational_operators/HashJoinOperator.cpp [hunk 8]
```diff
@@ -277,52 +377,59 @@ bool HashJoinOperator::getAllNonOuterJoinWorkOrders(
...
+        if (probe_relation_.hasNUMAPlacementScheme()) {
+          addPartitionAwareWorkOrders<JoinWorkOrderClass>(container,
+                                                          query_context,
+                                                          storage_manager,
+                                                          residual_predicate,
+                                                          selection,
+                                                          output_destination);
         }
-        started_ = true;
+#endif
```
Comment (hbdeshmukh) [inline relational_operators/HashJoinOperator.cpp]: Do we consider a case when the relation is partitioned but is not NUMA partitioned?

Comment (zuyu) [inline relational_operators/HashJoinOperator.cpp]: We should, i.e., in the distributed case.

**Comment by adalbertgerald** [comment]
Thanks for your comments @zuyu and @harshad. I have updated the buildhash and the hashjoin operators. Can you please take a look?

File: query_optimizer/ExecutionGenerator.cpp [hunk 9]
```diff
@@ -670,10 +670,22 @@ void ExecutionGenerator::convertHashJoin(const P::HashJoinPtr &physical_plan) {
...
+  if (build_relation_info->relation->hasPartitionScheme()) {
+    const std::size_t num_partitions =
+        build_relation_info->relation->getPartitionScheme().getPartitionSchemeHeader().getNumPartitions();
+    join_hash_table_indices.resize(num_partitions);
+  } else {
+    join_hash_table_indices.resize(1);
+  }
+
+  for (int i = 0; i < join_hash_table_indices.size(); ++i) {
+    join_hash_table_indices[i] = query_context_proto_->join_hash_tables_size();
```
Comment (zuyu) [inline query_optimizer/ExecutionGenerator.cpp]: If I understand correctly, with partition, we should create one hash table (proto) per partition. But here I believe it is a bug that we copy the only hash table for various partitions.

**Comment by adalbertgerald** [comment]
Hi @hbdeshmukh, I have made the changes for partition-aware hash joins. Can you please take a look?

File: relational_operators/WorkOrder.proto [hunk 10]
```diff
@@ -120,14 +122,15 @@ message HashJoinWorkOrder {
     repeated int32 join_key_attributes = 163;
     optional bool any_join_key_attributes_nullable = 164;
     optional int32 insert_destination_index = 165;
-    optional uint32 join_hash_table_index = 166;
-    optional int32 selection_index = 167;
-    optional fixed64 block_id = 168;
+    optional uint32 join_hash_table_group_index = 166;
+    optional uint32 join_hash_table_index = 167;
```
Comment (zuyu) [inline relational_operators/WorkOrder.proto]: Q: what's the difference between `join_hash_table_group_index` and `join_hash_table_index`.

File: relational_operators/DestroyHashOperator.hpp [hunk 11]
```diff
@@ -75,20 +89,28 @@ class DestroyHashWorkOrder : public WorkOrder {
...
    *
-   * @param hash_table_index The index of the JoinHashTable in QueryContext.
+   * @param hash_table_group_index The group index of the JoinHashTable in QueryContext.
    * @param query_context The QueryContext to use.
    **/
-  DestroyHashWorkOrder(const QueryContext::join_hash_table_id hash_table_index,
-                       QueryContext *query_context)
-      : hash_table_index_(hash_table_index),
-        query_context_(DCHECK_NOTNULL(query_context)) {}
+  DestroyHashWorkOrder(const CatalogRelation &input_relation,
```
Comment (zuyu) [inline relational_operators/DestroyHashOperator.hpp]: @adalbertgerald We could not use `CatalogRelation` in any `WorkOrder`s. To get the partition info, you need to put it in `WorkOrder` proto.

File: relational_operators/WorkOrder.proto (hunk 10 above)
Comment (adalbertgerald) [inline relational_operators/WorkOrder.proto]: Now, there may be a group of hash tables for each build relation (one for each partition). 'join_hash_table_group_index' refers to a group of hash tables and 'join_hash_table_index' refers to a single hash table within that group.

File: relational_operators/DestroyHashOperator.hpp (hunk 11 above)
Comment (adalbertgerald) [inline relational_operators/DestroyHashOperator.hpp]: Do you mean we should have the relation id in DestroyHashWorkOrder proto and retrieve the CatalogRelation from it? Or do you mean something else?

Comment (zuyu) [inline relational_operators/DestroyHashOperator.hpp]: Regarding `input_relation_`, we cannot access `CatalogRelation` in `WorkOrder` at all, only `CatalogRelationSchema`.

File: relational_operators/WorkOrder.proto (hunk 10 above)
Comment (zuyu) [inline relational_operators/WorkOrder.proto]: Do you have comments somewhere in the code?

File: relational_operators/WorkOrderFactory.cpp [hunk 12]
```diff
@@ -113,7 +113,8 @@ WorkOrder* WorkOrderFactory::ReconstructFromProto(const serialization::WorkOrder
     case serialization::DESTROY_HASH: {
       LOG(INFO) << "Creating DestroyHashWorkOrder";
       return new DestroyHashWorkOrder(
-          proto.GetExtension(serialization::DestroyHashWorkOrder::join_hash_table_index),
+          *(catalog_database->getRelationById(proto.GetExtension(serialization::DestroyHashWorkOrder::relation_id))),
```
Comment (zuyu) [inline relational_operators/WorkOrderFactory.cpp]: Use `getRelationSchemaById`.

File: query_execution/QueryContext.proto [hunk 13]
```diff
@@ -42,15 +42,20 @@ message QueryContext {
...
-  repeated SortConfiguration sort_configs = 8;
-  repeated Tuple tuples = 9;
+  repeated HashTableGroup join_hash_table_groups = 2;
+  repeated InsertDestination insert_destinations = 3;
+  repeated Predicate predicates = 4;
+  repeated ScalarGroup scalar_groups = 5;
+  repeated SortConfiguration sort_configs = 6;
+  repeated Tuple tuples = 7;
+  repeated GeneratorFunctionHandle generator_functions = 8;
+  repeated BloomFilter bloom_filters = 9;
```
Comment (zuyu) [inline query_execution/QueryContext.proto]: I don't understand why we move `BloomFilter` and `GeneratorFunctionHandle` to the end? We should keep all elements in the alphabet order.

File: query_optimizer/CMakeLists.txt [hunk 14]
```diff
@@ -1,4 +1,4 @@
-#   Copyright 2011-2015 Quickstep Technologies LLC.
+#   Copyright 2013-2015 Quickstep Technologies LLC.
```
Comment (zuyu) [inline query_optimizer/CMakeLists.txt]: I don't understand why we changed the year.

File: catalog/CatalogDatabaseLite.hpp [hunk 15]
```diff
@@ -69,6 +70,16 @@ class CatalogDatabaseLite {
   virtual const CatalogRelationSchema& getRelationSchemaById(const relation_id id) const = 0;
 
   /**
+   * @brief Get a relation by ID.
+   *
+   * @param id The id to search for.
+   * @return The relation with the given ID.
+   **/
+  virtual const CatalogRelation* getRelationById(const relation_id id) const {
```
Comment (zuyu) [inline catalog/CatalogDatabaseLite.hpp]: Remove this one. We don't need it in the base class.

File: relational_operators/WorkOrderFactory.cpp [hunk 16]
```diff
@@ -453,9 +455,9 @@ bool WorkOrderFactory::ProtoIsValid(const serialization::WorkOrder &proto,
 
       return proto.HasExtension(serialization::BuildHashWorkOrder::any_join_key_attributes_nullable) &&
              proto.HasExtension(serialization::BuildHashWorkOrder::block_id) &&
-             proto.HasExtension(serialization::BuildHashWorkOrder::join_hash_table_index) &&
+             proto.HasExtension(serialization::BuildHashWorkOrder::join_hash_table_group_index) &&
              query_context.isValidJoinHashTableId(
-                 proto.GetExtension(serialization::BuildHashWorkOrder::join_hash_table_index));
+                 proto.GetExtension(serialization::BuildHashWorkOrder::join_hash_table_group_index));
```
Comment (zuyu) [inline relational_operators/WorkOrderFactory.cpp]: We should keep the checks for `join_hash_table_index`, and add new ones for `join_hash_table_group_index`.

File: relational_operators/WorkOrder.proto [hunk 17]
```diff
@@ -83,7 +84,8 @@ message DeleteWorkOrder {
 message DestroyHashWorkOrder {
   extend WorkOrder {
     // All required.
-    optional uint32 join_hash_table_index = 112;
+    optional int32 relation_id = 112;
```
Comment (zuyu) [inline relational_operators/WorkOrder.proto]: We should serialize `PartitionSchemeHeader` instead of `relation_id`. Also, this entry is not required, so move it after `join_hash_table_group_index`.

File: relational_operators/DestroyHashOperator.hpp (hunk 11 above)
Comment (adalbertgerald) [inline relational_operators/DestroyHashOperator.hpp]: But we need input_relation to find the number of partitions. Do you have any ideas about how we could do this?

File: relational_operators/WorkOrderFactory.cpp (hunk 12 above)
Comment (adalbertgerald) [inline relational_operators/WorkOrderFactory.cpp]: We don't need the schema. We need the relation.

File: query_optimizer/CMakeLists.txt (hunk 14 above)
Comment (adalbertgerald) [inline query_optimizer/CMakeLists.txt]: I too don't understand how it got changed! :) Correcting it now.

File: query_execution/QueryContext.proto (hunk 13 above)
Comment (adalbertgerald) [inline query_execution/QueryContext.proto]: I too don't understand how it got changed! :) Correcting it now.

File: relational_operators/HashJoinOperator.hpp [hunk 18]
```diff
@@ -308,7 +376,11 @@ class HashInnerJoinWorkOrder : public WorkOrder {
         selection_(selection),
         hash_table_(hash_table),
         output_destination_(DCHECK_NOTNULL(output_destination)),
-        storage_manager_(DCHECK_NOTNULL(storage_manager)) {}
+        storage_manager_(DCHECK_NOTNULL(storage_manager)) {
+    if (numa_node != -1) {
+      preferred_numa_nodes_.push_back(numa_node);
+    }
```
Comment (zuyu) [inline relational_operators/HashJoinOperator.hpp]: I would suggest to add an explicit `WorkOrder` constructor:

```
  // WorkOrder.hpp
  explicit WorkOrder(const numa_node_id numa_node = -1) {
    if (numa_node != -1) {
      preferred_numa_nodes_.push_back(numa_node);
    }
  }

  // HashJoinOperator.hpp
  HashInnerJoinWorkOrder(...)
      : WorkOrder(numa_node)
        ... {}
```

File: catalog/CatalogDatabaseLite.hpp (hunk 15 above)
Comment (adalbertgerald) [inline catalog/CatalogDatabaseLite.hpp]: Good catch. Getting rid of it now.

File: relational_operators/SelectOperator.cpp [hunk 19]
```diff
@@ -168,17 +168,13 @@ bool SelectOperator::getAllWorkOrders(
 }
 
 void SelectWorkOrder::execute() {
-  BlockReference block(
-      storage_manager_->getBlock(input_block_id_, input_relation_, getPreferredNUMANodes()[0]));
+  BlockReference block(storage_manager_->getBlock(
+      input_block_id_, input_relation_, getPreferredNUMANodes().empty() ? -1 : getPreferredNUMANodes()[0]));
```
Comment (zuyu) [inline relational_operators/SelectOperator.cpp]: Just using `preferred_numa_nodes_` is ok.

File: relational_operators/SelectOperator.cpp [hunk 20]
```diff
@@ -168,17 +168,13 @@ bool SelectOperator::getAllWorkOrders(
...
   if (simple_projection_) {
-    block->selectSimple(simple_selection_,
-                        predicate_,
-                        output_destination_);
+    block->selectSimple(simple_selection_, predicate_, output_destination_);
   } else {
-    block->select(*DCHECK_NOTNULL(selection_),
-                  predicate_,
-                  output_destination_);
+    block->select(*DCHECK_NOTNULL(selection_), predicate_, output_destination_);
```
Comment (zuyu) [inline relational_operators/SelectOperator.cpp]: These changes are not related to this PR. Please reset them.

File: query_execution/QueryContext.hpp [hunk 21]
```diff
@@ -302,8 +307,8 @@ class QueryContext {
    *
    * @return True if valid, otherwise false.
    **/
-  bool isValidJoinHashTableId(const join_hash_table_id id) const {
-    return id < join_hash_tables_.size();
+  bool isValidJoinHashTableId(const join_hash_table_group_id gid, const join_hash_table_id id = 0) const {
```
Comment (zuyu) [inline query_execution/QueryContext.hpp]: Please don't use a default value here. Also, please update the `doxygen` comments for updated arguments.

File: query_execution/QueryContext.hpp [hunk 22]
```diff
@@ -313,19 +318,19 @@ class QueryContext {
    *
    * @return The JoinHashTable, already created in the constructor.
    **/
-  inline JoinHashTable* getJoinHashTable(const join_hash_table_id id) {
-    DCHECK_LT(id, join_hash_tables_.size());
-    return join_hash_tables_[id].get();
+  inline JoinHashTable* getJoinHashTable(const join_hash_table_group_id gid, const join_hash_table_id id = 0) {
+    DCHECK_LT(id, join_hash_tables_[gid].size());
```
Comment (zuyu) [inline query_execution/QueryContext.hpp]: Please check `gid` is valid before using it to access the array.

File: query_execution/QueryContext.hpp [hunk 23]
```diff
@@ -313,19 +318,19 @@ class QueryContext {
...
   /**
    * @brief Destory the given JoinHashTable.
    *
    * @param id The id of the JoinHashTable to destroy.
    **/
-  inline void destroyJoinHashTable(const join_hash_table_id id) {
-    DCHECK_LT(id, join_hash_tables_.size());
-    join_hash_tables_[id].reset();
+  inline void destroyJoinHashTable(const join_hash_table_group_id gid, const join_hash_table_id id = 0) {
+    DCHECK_LT(id, join_hash_tables_[gid].size());
```
Comment (zuyu) [inline query_execution/QueryContext.hpp]: Please check `gid` is valid before using it to access the array.

File: query_execution/QueryContext.hpp [hunk 24]
```diff
@@ -465,7 +470,7 @@ class QueryContext {
   std::vector<std::unique_ptr<BloomFilter>> bloom_filters_;
   std::vector<std::unique_ptr<const GeneratorFunctionHandle>> generator_functions_;
   std::vector<std::unique_ptr<InsertDestination>> insert_destinations_;
-  std::vector<std::unique_ptr<JoinHashTable>> join_hash_tables_;
+  std::vector<std::vector<std::unique_ptr<JoinHashTable>>> join_hash_tables_;
```
Comment (zuyu) [inline query_execution/QueryContext.hpp]: Please rename to `join_hash_table_groups_`.

File: query_optimizer/ExecutionGenerator.cpp [hunk 25]
```diff
@@ -714,25 +716,69 @@ void ExecutionGenerator::convertHashJoin(const P::HashJoinPtr &physical_plan) {
...
-          key_types));
-
-  const CatalogRelationSchema *build_relation = build_relation_info->relation;
-  for (const attribute_id build_attribute : build_attribute_ids) {
-    hash_table_proto->add_key_types()->CopyFrom(
-        build_relation->getAttributeById(build_attribute)->getType().getProto());
-  }
+  const QueryContext::join_hash_table_group_id join_hash_table_group_index =
+      query_context_proto_->join_hash_table_groups_size();
+  query_context_proto_->add_join_hash_table_groups();
```
Comment (zuyu) [inline query_optimizer/ExecutionGenerator.cpp]: Please save the pointer here, and do not later use `query_context_proto_->mutable_join_hash_table_groups(join_hash_table_group_index)` to get the pointer.

File: query_optimizer/ExecutionGenerator.cpp [hunk 26]
```diff
@@ -714,25 +716,69 @@ void ExecutionGenerator::convertHashJoin(const P::HashJoinPtr &physical_plan) {
...
+  std::vector<S::HashTable *> hash_table_proto;
```
Comment (zuyu) [inline query_optimizer/ExecutionGenerator.cpp]: You don't actually need this local variable. Save the pointer only when you use it.

File: query_optimizer/ExecutionGenerator.cpp [hunk 27]
```diff
@@ -714,25 +716,69 @@ void ExecutionGenerator::convertHashJoin(const P::HashJoinPtr &physical_plan) {
...
+                build_relation->getAttributeById(build_attribute)->getType().getProto());
+          }
+
+          hash_table_proto[part_id]->set_estimated_num_entries(build_cardinality);
+        }
+      }
+    }
+  } else {
+    hash_table_proto.emplace_back(
+        query_context_proto_->mutable_join_hash_table_groups(join_hash_table_group_index)->add_join_hash_tables());
```
Comment (zuyu) [inline query_optimizer/ExecutionGenerator.cpp]: Just save the pointer, and remove `hash_table_proto`.

**Comment by zuyu** [comment]
@adalbertgerald I believe there is a fundamental design issue that we tie the partition info to close with NUMA support. So in future we would extend the partition info to distributed workers (but no NUMA support), it needs significant rework.

Could you please come up with an approach that first check wether we have partition info, and then use different strategies to achieve partitioning. In this case, we do in the NUMA case. And in future, we may only use the distributed case. Thanks!

File: catalog/CatalogDatabaseLite.hpp (hunk 15 above)
Comment (adalbertgerald) [inline catalog/CatalogDatabaseLite.hpp]: This is needed else WorkOrderFactory will not compile.

File: relational_operators/DestroyHashOperator.hpp (hunk 11 above)
Comment (zuyu) [inline relational_operators/DestroyHashOperator.hpp]: Serialize the whole `PartitionSchemaHeader`.

File: relational_operators/WorkOrderFactory.cpp (hunk 12 above)
Comment (zuyu) [inline relational_operators/WorkOrderFactory.cpp]: We need `PartitionSchemeHeader`, not `CatalogRelation`. Serialize it in `WorkOrder` proto.

File: catalog/CatalogDatabaseLite.hpp (hunk 15 above)
Comment (zuyu) [inline catalog/CatalogDatabaseLite.hpp]: This is an incorrect design. `WorkOrderFactory` should not use it at all.

**Comment by adalbertgerald** [comment]
Reopened a new pull request #238.
